### ISSUE TITLE
REST API & Static Command Services enhancements + GridView and DataPager client-only mode

### ISSUE THREAD

**Issue by tomasherceg** [riganti/dotvvm#605]
In order to support `GridViewDataSet` in REST API bindings, we need to add the "client-only" mode so the controls don't need to make a postback to change the sort or paging options.

The problem is that when I use REST API binding (or a Static Command Service) on the server, they return `null`, so the controls don't render anything, however when I change the sort column or click on a link in pager, the controls require a proper control tree (but they don't see any items).

We should add `ClientOnly` property in `GridView` and `DataPager`. 
(the name of the property can be different - e.g. `AllowClientOnlyData`, or alternatively, we can create `ClientGridView` and `ClientDataPager` controls)

* The controls won't build the control tree for their events, they will even not evaluate the `DataSource`.
* Command bindings should not be allowed in their templates.
* When the user clicks the column header, the `SortingOptions` in `GridViewDataSet` will be changed and `IsRefreshRequested` will be set to `true`.
* When the user clicks the column header, the `PagingOptions` in `GridViewDataSet` will be changed and `IsRefreshRequested` will be set to `true`.

The following scenarios should be supported:

1. Loading data from a static command service or a REST API method and auto-refreshing whenever sorting or paging data are changed (so we don't need to use `_api.RefreshOnChange(..., DataSet.IsRefreshRequired`):

`<dot:GridView DataSource="{value: _myApi.GetCountries()}">...`

2. Some kind of assignment, so we can store the dataset in the viewmodel and use it in other controls - ideally with `Bind(Direction.ServerToClientFirstRequest)` allowing to set default page size etc.

```
<dot:GridView DataSource="{value: DataSet = _myApi.GetCountries()}">
   ...
</dot:GridView>
<dot:DataPager DataSet="{value: _myApi.GetCountries()}">
```

It doesn't have to be an assignment, it can be something like `_api.Store(value, target)`, `_api.Cache(value, target)`.

The same should work with static command services, so they can be used instead of REST API bindings. Also, we should make `_api.RefreshOn...` work with them.

**Comment by exyi** [comment]
There is a few fundamental problems:
* The `GridViewDataSet` contains both the data and the configuration required to load it, which is usable using the "old way" of loading data, but it implies that it can't be loaded from API as a whole.
* Both GridView and DataPager use command bindings extensively to change the configuration in GridViewDataSet and reload the data.


We could split the data and the configuration into separate properties, but this would require to redesign all the that use GridViewDataSet for paging/sorting/...


I propose creating a data set that would handle the loading client-side, something like ClientLoadingDataSet. It would require a Linq.Expression in a constructor that would be translated to Javascript and would be executed client-side every time some property affecting the loading changes.
The loading expression would get reference to the DataSet (for the paging/sorting/... info) and some context that will be used to access extension parameters ("_api", ...). It could be initialized like this:

```C#
ClientLoadingDataSet<SomeObject> Data = new ClientLoadingDataSet<SomeObject>(
    load: (dataSet, context) => context.ExtensionParameter<MyApiClient>("myApi").LoadSomething(dataSet.PageSize, dataSet.PageIndex)
);
```
or like this if you'd like to invoke static command:

```C#
ClientLoadingDataSet<SomeObject> Data = new ClientLoadingDataSet<SomeObject>(
    load: (dataSet, context) => context.Service<SomeLoaderService>().LoadSomething(dataSet.PagingOptions)
);
```

The data set will still have the items in it, but it will never send them to server as they will be marked as `[Bind(ServerToClientFirstRequest)]`. It would allow preloading for first request and load client-side asynchronously on every refresh.

We would also have to tweak the GridView and DataPager control to suppress the command when it's enough to change page/sort/... client side. We could either have the `AllowClientOnlyData` property or decide which mode to use from the result type of `DataSource` binding. It should however also work with controls unaware of this new client-side DataSet, as it would just change the properties on server (so it's inefficient, but will work).

**Comment by tomasherceg** [comment]
Agreed. Also, since the `GridView` and `DataPager` should generate different command bindings for sorting and paging based on dataset types, I would like to have some strategies for classic and client-side dataset types - something like "command providers" with methods for creating sort command for a button in GridView header and commands for buttons in DataPager. We may add another types of datasets in the future.

**Comment by adamjez** [comment]
@exyi I have a few issues with your approach. 

```CSHARP
ClientLoadingDataSet<SomeObject> Data = new ClientLoadingDataSet<SomeObject>(
    load: (dataSet, context) => context.ExtensionParameter<MyApiClient>("myApi").LoadSomething(dataSet.PageSize, dataSet.PageIndex)
);
```

When you are providing an expression which loads data to client-side DataSet, you can't refer to other **ViewModel** properties. There is no **DataContext** and it can't be created, because the DataSet can be used in multiple different DataContexts. Even if you provide current viewModel type as a DataContext, you won't be able to refer to `_parent` or `_root`.
